File: 14.最长公共前缀.java
/*
 * @lc app=leetcode.cn id=14 lang=java
 *
 * [14] 最长公共前缀
 */

// @lc code=start
class Solution {
    public String longestCommonPrefix(String[] strs) {
        if(strs.length == 0) return "";
        int minLen = strs[0].length();
        for(int i = 1; i < strs.length; i ++){
            minLen = Math.min(minLen, strs[i].length());
        }
        int l = 0, r = minLen - 1;
        while(l <= r){
            int middle = (r + l) / 2;
            if(isPrefixMatch(strs, middle)){
                l = middle + 1;
            }else{
                r = middle - 1;
            }
        }
        return strs[0].substring(0, (r + l) / 2);

    }

    public boolean isPrefixMatch(String[] strs, int len){
        String prefix = strs[0].substring(0, len + 1);
        for(int i = 1; i < strs.length; i ++){
            if(!strs[i].startsWith(prefix)) return false;
        }
        return true;
    }
}
// @lc code=end

File: 14.最长公共前缀.go
/*
 * @lc app=leetcode.cn id=14 lang=golang
 *
 * [14] 最长公共前缀
 */

// @lc code=start
func longestCommonPrefix(strs []string) string {
	if strs == nil || len(strs) == 0 {
		return ""
	}

	var length = len(strs[0])
	for i := 0; i < length; i++ {
		var c = strs[0][i]
		for j := 1; j < len(strs); j++ {
			if len(strs[j]) < i + 1 || c != strs[j][i] {
				return strs[0][:i]
			}
		}
	}
	return strs[0]
}
// @lc code=end

